lib/bridge.js

```diff
@@ -773,61 +776,92 @@ Bridge.prototype._onEvent = function(event) {
         }
     }
 
-    var request = this._requestFactory.newRequest({ data: event });
-    var context = new BridgeContext({
-        sender: event.user_id,
-        target: event.state_key,
-        room: event.room_id
-    });
-    var data = {
+    const request = this._requestFactory.newRequest({ data: event });
+    const contextReady = this._getBridgeContext(event);
+    const dataReady = contextReady.then(c => ({
         request: request,
-        context: context
-    };
+        context: c
+    }));
 
-    var promise;
-    if (this.opts.disableContext) {
-        promise = Promise.resolve();
-        data.context = null;
-    }
-    else {
-        promise = context.get(this._roomStore, this._userStore);
-    }
+    const dataReadyLimited = this._limited(dataReady, request);
 
-    if (this.opts.queue.type === "none") { // consume as soon as we have context
-        promise.done(() => {
-            this._onConsume(null, data);
-        }, (err) => {
-            this._onConsume(err);
-        });
-        return request.getPromise();
-    }
+    this._queue.push(event, dataReadyLimited);
+    this._queue.consume();
+
+    return (
+        request
+        .getPromise()
+        .catch(EventNotHandledError, e => this._handleEventError(event, e))
+    );
+};
 
+/**
+ * Returns a promise that is like `promise` but adheres to all necessary limits.
+ */
+Bridge.prototype._limited = function(promise, request) {
+    // queue.perRequest controls whether multiple request can be processed by
+    // the bridge at once.
     if (this.opts.queue.perRequest) {
-        promise = Promise.settle([
-            promise,
-            this._prevRequestPromise
-        ]);
+        const promiseLimited = this._prevRequestPromise.reflect().return(promise);
         this._prevRequestPromise = request.getPromise();
+        return promiseLimited;
     }
 
-    this._queue.push(event, data, promise);
-    this._queue.consume();
-    return request.getPromise();
-};
+    return promise;
+}
 
 Bridge.prototype._onConsume = function(err, data) {
-    if (!err) {
-        this.opts.controller.onEvent(data.request, data.context);
+    if (err) {
+        // The data for the event could not be retrieved.
+        // In this case we reject the request and the event will be delivered
+        // by the HS to the bridge again.
+        data.request.reject();
+        this.opts.controller.onLog("onEvent failure: " + err);
         return;
     }
-    if (!this.opts.controller.onLog) {
-        return;
+
+    this.opts.controller.onEvent(data.request, data.context);
+};
+
+Bridge.prototype._getBridgeContext = function(event) {
```

Comment:
async function?

Reply:
Well, it does return a Promise… If I added the `asyc` keyword, wouldn't that make it return `Promise<Promise<BridgeContext|null>>`?

Reply:
JS is smart, and it doesn't wrap promises like that but chains them (so you will never have to handle a promise inside a promise). Making the func `async` means you can `return null;` rather than `return Promise.resolve(null);`